### PR TITLE
feat: add ApiDefinitionAnnotationProcessor

### PR DESCRIPTION
**Pull request by Fox32**
This is a PoC for #2834 

#### :heavy_check_mark: Checklist

- [ ] All tests are passing `yarn test`
- [ ] Screenshots attached (for UI changes)
- [ ] Relevant documentation updated
- [ ] Prettier run on changed files
- [ ] Tests added for new functionality
- [ ] Regression tests added for bug fixes
- [ ] Added a changeset ([more info](https://github.com/spotify/backstage/blob/master/CONTRIBUTING.md#creating-changesets))

### CODE REVIEW

File: plugins/catalog-backend/src/ingestion/processors/ApiDefinitionAnnotationProcessor.ts [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+  }
+
+  return Object.keys(entity.metadata.annotations).filter(a =>
+    a.match(API_DEFINITION_TYPE_ANNOTATION),
+  );
+}
+
+export function extractApiDefinitionType(annotationName: string): string {
+  return annotationName.match(API_DEFINITION_TYPE_ANNOTATION)![1];
+}
```
Comment (freben) [inline plugins/catalog-backend/src/ingestion/processors/ApiDefinitionAnnotationProcessor.ts:38]: These two could mayhap be a single `getApiDefinitionUrls(entity: Entity): Record<string, URL>`, returning something like

```js
{
  openapi: new URL('https://example.com/openapi.yaml')
}
```

where the URL is always absolute. If the input annotation was relative, you'd make that into an absolute one relative to the location that the entity itself is located on.

File: plugins/catalog-backend/src/ingestion/processors/ApiDefinitionAnnotationProcessor.ts [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+
+export class ApiDefinitionAnnotationProcessor implements LocationProcessor {
+  async processEntity(
+    entity: Entity,
+    location: LocationSpec,
+    emit: LocationProcessorEmit,
+  ): Promise<Entity> {
+    // TODO: Case insensitive?
```
Comment (freben) [inline plugins/catalog-backend/src/ingestion/processors/ApiDefinitionAnnotationProcessor.ts:46]: I guess, yeah... Everything else tries to be, internally, but then again we kinda expect users to write it exactly like that in yaml.

File: plugins/catalog-backend/src/ingestion/processors/ApiDefinitionAnnotationProcessor.ts [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+  ): Promise<ComponentEntity> {
+    const url = entity.metadata.annotations![annotation];
+    const type = extractApiDefinitionType(annotation);
+    const name = `${entity.metadata.name}-${type}-api`;
+    const apiEntity = {
+      apiVersion: 'backstage.io/v1alpha1',
+      kind: 'API',
+      metadata: {
+        name,
+        description: entity.metadata.description,
```
Comment (freben) [inline plugins/catalog-backend/src/ingestion/processors/ApiDefinitionAnnotationProcessor.ts:84]: isn't this something that can be found in the api specification?

File: plugins/catalog-backend/src/ingestion/processors/ApiDefinitionAnnotationProcessor.ts [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+    };
+
+    emit({
+      type: 'entity',
+      entity: apiEntity,
+      location,
+    });
+
+    if (!entity.spec.implementsApis) {
+      entity.spec.implementsApis = [name];
```
Comment (freben) [inline plugins/catalog-backend/src/ingestion/processors/ApiDefinitionAnnotationProcessor.ts:103]: Yeah i know you've seen the current ongoing work in relations. Currently we are modifying the actual spec, but we're pondering whether we really should - consumers are expected to use the relations as the authoritative source anyway in the end.

Welp, just a side note - for now, what you do here is fine.

**Review by freben — COMMENTED** [review]
Hey, this has been on my shame list of postponement for the longest time. Gave some rough feedback.

So yeah this is definitely possible. We had some discussions too about whether this should be an annotation or not, seeing as you may want to implement several apis and may want to define some metadata for each (like the generated name or so). What are your current thoughts / plans regarding this?

**Comment by freben** [comment]
Another possibility:

```yaml
spec:
  implementsApis:
    - location: ./openapi.spec
      name: my-openapi-thing
      ...etc
```

File: plugins/catalog-backend/src/ingestion/processors/ApiDefinitionAnnotationProcessor.ts (hunk 3 above)
Comment (Fox32) [reply plugins/catalog-backend/src/ingestion/processors/ApiDefinitionAnnotationProcessor.ts:84]: Yes, depending on the format we might find a title there (asyncapi, openapi), but for other format we don't have it (like grpc).

I would like to add a second processor that can extract the title from the specs that support that. But in a separate PR.

**Comment by Fox32** [comment]
@freben Thanks for your feedback

Yeah, after the discussions about annotation, the variant with the complex `implementsApis` entries feels appealing. I'm just wondering how to document this. I would keep the spec (validation) of the component as-is, but allow users to add complex objects there. The processor evaluates them and replaces them with the simpler links to the API entity. So only the simple links are stored. And here is probably were the relation system comes into play. Therefore it's fine for me to postpone this for a week and see how the relations come along.

**Comment by freben** [comment]
Exactly, that does beg the question, should one really rewrite the spec at all. It sounds more appealing to eat the format as is and spit out relations and tell people to read those instead of consuming the spec.

**Comment by Fox32** [comment]
I will close this for now, but might revisit this later

**Comment by andrewthauer** [comment]
For reference we built a somewhat similar processor to add API entities automatically. Except the difference is it does a lookup for a published API spec at a known location. This means it doesn't require any changes in provides by component catalog file. We don't have a consumes equivalent yet though.

**Pull request closed without merging**